File: library/src/main/java/me/proxer/library/api/EpisodeInfoAdapter.java
package me.proxer.library.api;

import com.squareup.moshi.FromJson;
import com.squareup.moshi.Json;
import com.squareup.moshi.ToJson;
import me.proxer.library.api.EpisodeInfoAdapter.IntermediateEpisodeInfo.IntermediateEpisode;
import me.proxer.library.entity.info.AnimeEpisode;
import me.proxer.library.entity.info.Episode;
import me.proxer.library.entity.info.EpisodeInfo;
import me.proxer.library.entity.info.MangaEpisode;
import me.proxer.library.enums.Category;
import me.proxer.library.enums.MediaLanguage;

import javax.annotation.Nullable;
import java.util.*;

/**
 * @author Ruben Gees
 */
class EpisodeInfoAdapter {

    private static final String DELIMITER = ",";

    @FromJson
    EpisodeInfo fromJson(final IntermediateEpisodeInfo json) {
        final List<Episode> episodes = new ArrayList<>(json.episodes.size());

        for (IntermediateEpisode episode : json.episodes) {
            if (json.category == Category.ANIME) {
                episodes.add(new AnimeEpisode(episode.number, episode.language,
                        new HashSet<>(Arrays.asList(episode.hosters.split(DELIMITER))),
                        Arrays.asList(episode.hosterImages.split(DELIMITER))));
            } else {
                episodes.add(new MangaEpisode(episode.number, episode.language, episode.title));
            }
        }

        return new EpisodeInfo(json.firstEpisode, json.lastEpisode, json.category, json.availableLanguages,
                json.userState, episodes);
    }

    @SuppressWarnings("unused")
    @ToJson
    String toJson(@SuppressWarnings("unused") @Nullable final EpisodeInfo episodeInfo) {
        // Moshi somehow needs this method for this case, we don't use it though.
        throw new UnsupportedOperationException();
    }

    static class IntermediateEpisodeInfo {

        @Json(name = "start")
        int firstEpisode;

        @Json(name = "end")
        int lastEpisode;

        @Json(name = "kat")
        Category category;

        @Json(name = "lang")
        Set<MediaLanguage> availableLanguages;

        @Json(name = "state")
        int userState;

        @Json(name = "episodes")
        List<IntermediateEpisode> episodes;

        static class IntermediateEpisode {

            @Json(name = "no")
            int number;

            @Json(name = "typ")
            MediaLanguage language;

            @Json(name = "title")
            String title;

            @Json(name = "types")
            String hosters;

            @Json(name = "typeimg")
            String hosterImages;
        }
    }
}

File: library/src/main/kotlin/me/proxer/library/api/EpisodeInfoAdapter.kt
package me.proxer.library.api

import com.squareup.moshi.FromJson
import com.squareup.moshi.Json
import com.squareup.moshi.JsonClass
import com.squareup.moshi.ToJson
import me.proxer.library.api.EpisodeInfoAdapter.IntermediateEpisodeInfo.IntermediateEpisode
import me.proxer.library.entity.info.AnimeEpisode
import me.proxer.library.entity.info.EpisodeInfo
import me.proxer.library.entity.info.MangaEpisode
import me.proxer.library.enums.Category
import me.proxer.library.enums.MediaLanguage

/**
 * @author Ruben Gees
 */
internal class EpisodeInfoAdapter {

    private companion object {
        private const val DELIMITER = ","
    }

    @FromJson
    fun fromJson(json: IntermediateEpisodeInfo): EpisodeInfo {
        val episodes = when (json.category) {
            Category.ANIME -> json.episodes.map { episode ->
                val hosters = episode.hosters?.split(DELIMITER)?.toSet()
                    ?: throw IllegalArgumentException("episode.hosters is null")

                val hosterImages = episode.hosterImages?.split(DELIMITER)
                    ?: throw IllegalArgumentException("episode.hosters is null")

                AnimeEpisode(episode.number, episode.language, hosters, hosterImages)
            }
            Category.MANGA -> json.episodes.map { episode ->
                val title = episode.title ?: throw IllegalArgumentException("episode.title is null")

                MangaEpisode(episode.number, episode.language, title)
            }
        }

        return EpisodeInfo(
            json.firstEpisode, json.lastEpisode, json.category, json.availableLanguages,
            json.userProgress, episodes
        )
    }

    @ToJson
    fun toJson(value: EpisodeInfo?): IntermediateEpisodeInfo? {
        if (value == null) return null

        val episodes = value.episodes.map {
            when (it) {
                is AnimeEpisode -> {
                    val joinedHosters = it.hosters.joinToString(DELIMITER)
                    val joinedHosterImages = it.hosterImages.joinToString(DELIMITER)

                    IntermediateEpisode(it.number, it.language, null, joinedHosters, joinedHosterImages)
                }
                is MangaEpisode -> IntermediateEpisode(it.number, it.language, it.title, null, null)
                else -> throw IllegalArgumentException("Unknown Episode type: ${it.javaClass.name}")
            }
        }

        return IntermediateEpisodeInfo(
            value.firstEpisode, value.lastEpisode, value.category, value.availableLanguages,
            value.userProgress, episodes
        )
    }

    @JsonClass(generateAdapter = true)
    internal class IntermediateEpisodeInfo(
        @Json(name = "start") val firstEpisode: Int,
        @Json(name = "end") val lastEpisode: Int,
        @Json(name = "kat") val category: Category,
        @Json(name = "lang") val availableLanguages: Set<MediaLanguage>,
        @Json(name = "state") val userProgress: Int,
        @Json(name = "episodes") val episodes: List<IntermediateEpisode>
    ) {

        @JsonClass(generateAdapter = true)
        internal class IntermediateEpisode(
            @Json(name = "no") val number: Int,
            @Json(name = "typ") val language: MediaLanguage,
            @Json(name = "title") val title: String?,
            @Json(name = "types") val hosters: String?,
            @Json(name = "typeimg") val hosterImages: String?
        )
    }
}
